pkg/cache/cluster.go

```diff
@@ -813,6 +815,53 @@ func (c *clusterCache) IterateHierarchy(key kube.ResourceKey, action func(resour
 	}
 }
 
+// IterateGroupHierarchy executes multiple IterateHierarchy operations and only takes lock once.
+func (c *clusterCache) IterateGroupHierarchy(keys []kube.ResourceKey, action func(resource *Resource, namespaceResources map[kube.ResourceKey]*Resource)) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	var wg sync.WaitGroup
+	for _, key := range keys {
+		if res, ok := c.resources[key]; ok {
+			wg.Add(1)
+			go func() {
```

Comment:
Running separate goroutines for each resource is pretty expensive. I've tried to test this change on a large argocd instance and controller crashed.